### PR TITLE
Replace random nonce with offset

### PR DESCRIPTION
**Pull request by rklaehn**
This is pretty ugly, since I hacked in the offset all over the place instead of rethinking things a bit.

Will try rethinking this in this or a separate branch.

### CODE REVIEW

**Comment by rklaehn** [comment]
Interestingly, while chacha20 seek fails at values > 1<<38, *x* chacha20 works for the entire range of u64.

```rust
use chacha20::cipher::{NewCipher, StreamCipher, StreamCipherSeek};

fn chacha20_seek() {
    let key: chacha20::Key = [0u8; 32].into();
    let nonce: chacha20::Nonce = [0u8; 12].into();
    let mut cipher = chacha20::ChaCha20::new(&key, &nonce);
    cipher.seek(1u64<<37); // 38 fails
    let mut data = [1u8; 1024];
    cipher.apply_keystream(&mut data);
}

fn xchacha20_seek() {
    let key: chacha20::Key = [0u8; 32].into();
    let nonce: chacha20::XNonce = [0u8; 24].into();
    let mut cipher = chacha20::XChaCha20::new(&key, &nonce);
    cipher.seek(1u64<<62); // 38 fails
    let mut data = [1u8; 1024];
    cipher.apply_keystream(&mut data);
}
```

File: banyan/Cargo.toml [hunk 1]
```diff
@@ -18,7 +18,6 @@ futures = { version = "0.3.14", features = ["thread-pool"] }
 libipld = { version = "0.11.0", features = ["unleashed"] }
 maplit = "1.0.2"
 parking_lot = "0.11.1"
-rand = "0.8.3"
```
Comment (rklaehn) [inline banyan/Cargo.toml:21]: This makes me very happy...

**Review by dvc94ch — APPROVED** [review]
So what was the problem with the offsets being too small?

**Comment by rklaehn** [comment]
>  I guess I'm not particularly happy with xchacha happening to work but not chacha without understanding why.

Well, this is a draft. It would be good to know if this is a fundamental limitation of chacha or just a bug in the impl. If it was the latter I would prefer not to use xchacha.

File: banyan/src/zstd_dag_cbor_seq.rs [hunk 2]
```diff
@@ -248,42 +249,40 @@ impl ZstdDagCborSeq {
...
         // encrypt in place with the key and nonce
-        let mut chacha20 = XChaCha20::new(key, nonce);
+        let mut chacha20 = XChaCha20::new(key, &NONCE.into());
+        let len = data.len();
+        chacha20.seek(*offset);
         chacha20.apply_keystream(&mut data);
         // add the nonce
-        data.extend(nonce.as_slice());
+        data.extend(&offset.to_be_bytes());
+        *offset += len as u64;
```
Comment (rkuhn) [inline banyan/src/zstd_dag_cbor_seq.rs:266]: This seems to me like it should be the responsibility of the caller — unless `len` somehow is not known to upper layers (which wasn’t the impression I got).

File: banyan/src/zstd_dag_cbor_seq.rs [hunk 3]
```diff
@@ -248,42 +249,40 @@ impl ZstdDagCborSeq {
...
 
     /// decrypt using the given key
-    pub fn decrypt(data: &[u8], key: &chacha20::Key) -> anyhow::Result<Self> {
+    pub fn decrypt(data: &[u8], key: &chacha20::Key) -> anyhow::Result<(Self, u64)> {
         let (links, mut encrypted) = DagCborCodec.decode::<IpldNode>(data)?.into_data()?;
         let len = encrypted.len();
         anyhow::ensure!(len >= EXTRA_LEN);
-        let (compressed, nonce) = encrypted.split_at_mut(len - EXTRA_LEN);
-        XChaCha20::new(key, (&*nonce).into()).apply_keystream(compressed);
+        let (compressed, offset) = encrypted.split_at_mut(len - EXTRA_LEN);
```
Comment (rkuhn) [inline banyan/src/zstd_dag_cbor_seq.rs:277]: Why is this necessary? The offset should be known implicitly due to the position in the tree, right?

**Review by rkuhn — COMMENTED** [review]
So in essence this brings usage closer in line with the intended usage of the cipher, where the nonce comes from elsewhere. Why not pass it into the tree constructor? Client code can still pass in an array of zeros if they want.

One worry I have is that the nonce/offset combination should not be reused for different clear text (hence the “number used once” notion). This is obvious for the stream of payloads in the leaves, but it is not obvious for the stream of indexes in the branches — could you elaborate how this works?

File: banyan/src/zstd_dag_cbor_seq.rs (hunk 2 above)
Comment (rklaehn) [reply banyan/src/zstd_dag_cbor_seq.rs:266]: Well, this is just hacked in. I am working on a larger refactoring that will make things clearer. Basically, I really want to avoid having some misunderstanding that leads to offsets being reused.

File: banyan/src/zstd_dag_cbor_seq.rs (hunk 3 above)
Comment (rklaehn) [reply banyan/src/zstd_dag_cbor_seq.rs:277]: If you would never remove anything from the tree, you could infer the offset from the position in the tree. But we do want to have things like retain, which can forget parts of the tree. For that you need to explicitly store the offsets.

Now you could come up with a scheme where you only sometimes need the offset, but that would be rather complicated. Also, I prefer the blocks to be self-contained, so to look at a block you just need the secrets.

**Comment by rklaehn** [comment]
> So in essence this brings usage closer in line with the intended usage of the cipher, where the nonce comes from elsewhere. Why not pass it into the tree constructor? Client code can still pass in an array of zeros if they want.

For an event stream I don't see an advantage of making the nonce configurable. It would basically just be an additional piece of the secret key, but we do know that 32 bytes are enough. If you mess up handing out the offsets, making the nonce configurable won't help you - you will still have different data encrypted at the same offset, at which point your crypto is broken.

I was thinking of having a fun string in the nonce instead of just zeros. Some easter egg / inside joke...

> One worry I have is that the nonce/offset combination should must not be reused for different clear text (hence the “number used once” notion). This is obvious for the stream of payloads in the leaves, but it is not obvious for the stream of indexes in the branches — could you elaborate how this works?

Basically you have a per-stream counter that only ever increases, that gets used for both branches and leaves. When writing for a single stream (=same secret), you always feed from this counter for both leaves and branches. This is not that obvious in this version, but will become more obvious in the refactoring.

**Comment by rklaehn** [comment]
Changed the target to v13
